### DISCUSSION TITLE
Feature request: Timer

### DISCUSSION THREAD

**Discussion by wunderbaum** [dewenni/ESP32-Jarolift-Controller#4] [Ideas]
Implement timer for each cover.
With astro function according to longitude, latitude for sunset, sundown.

**Comment by dewenni** [comment]
Thank you for the function request. That sounds good and makes sense.   
Should also be feasible, but is also a bit of work in total.
lets see.

**Comment by dewenni** [comment]
Does this look good?

<img width="948" alt="image" src="https://github.com/user-attachments/assets/a6c93e9d-d696-4c7e-aa3b-1f853d8204c9" />

- fixes Time, Sunrise or Sundown
- optional offset for Sunrise, Sundown
- Command UP, DOWN
- selection of week days
- selection of shutters

Possibility to configure up to 6 such Timer?

**Reply by flushbug** [reply]
Looks great. Even though I will not be using it since I prefer to manage all the timers on one single place (home assistant)

**Reply by wunderbaum** [reply]
I really like this feature!
Of course, HA is the leading system. But I really appreciate it when systems are doing their own business without asking/waiting for the master.

**Comment by dewenni** [comment]
It is also not yet certain that I will realise this. All in all, it's a certain amount of work.
If there are people interested in this function, please let me know here

**Comment by dewenni** [comment]
yes, lets see. But I would simplify the configuration in the way that the user has to use the Bitmask like it is for the groups

<img width="852" alt="image" src="https://github.com/user-attachments/assets/9edabd43-b9de-4ce0-9054-9cccd6d8fa81" />

that saves me a lot of single web elements and variables to communicate between ESP and weblient.
Bitmask is only one variable for each (max 6) Timer instead of 16x6=96

**Comment by dewenni** [comment]
first implementation is done and seems to work so far:

<img width="977" alt="image" src="https://github.com/user-attachments/assets/fac58f84-6be5-46e8-a31f-5c46e69e6b54" />

I will do some further tests and build a new release in a couple of days

**Reply by flushbug** [reply]
What about a list of all available shutters with a checkbox after each to indicate that it should be considered for the timer ? Bitmask will be calculated automatically in the background based on this. I would also prefer this option for the group configuration. 

Would improve the user experience a lot.

**Reply by dewenni** [reply]
Yes, that could be a nice solution to have this as a Dialog that will open if you click on such a "Bitmask" field.
The Dialog shows all configured shutters with their configured names followed by a checkbox.
If you close the Dialog, the bit mask will be calculated and transferred to the input field.
That could be a good solution that does not need to have all the checkboxes multiple times for each Timer and Group.
Because this would increase the amount of HTML Code and also the communication between webUI and ESP a lot.
I will check how this could work...

**Comment by dewenni** [comment]
I have released the Timer feature with the new v1.5.0

> [!TIP]
> Remember that if you are already at v1.4.0, you can also use the possibility to update to this version right out of the WebUI of your controller if you click on the version information at the bottom left of the WebUI. [See also here](https://github.com/dewenni/ESP32-Jarolift-Controller?tab=readme-ov-file#github-ota-update)
